### PR TITLE
Source Code for Goldmine Algorithm in C++

### PR DESCRIPTION
**Pull request by weirdrag08**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

### Description

Adding Source Code for Goldmine Algorithm in C++

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

#2331

### CODE REVIEW

**Comment by weirdrag08** [comment]
Please review my code ;-)

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+// * For this algorithm we move column wise as our src as well as destination lies in first and last column respectively.
+// * There are 3 possible paths for each cell. One to top right , second to right , third to bottom right.
+// * Taking max out of all possible paths , we can earn maximum profit.
+
+class Solution
+{
+public:
+    int maxGold(int r, int c, vector<vector<int>> mine)
+    {
+        // * Code here
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:13]: Remove this unnecessary comments.

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        vector<vector<int>> dp(r, vector<int>(c));
+
+        for (int i = 0; i < r; i++)
+        {
+            for (int j = 0; j < c; j++)
+            {
+                if (j == 0)
+                {
+                    dp[i][j] = mine[i][j]; // * To fill the left wall
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:22]: Add comments above the line of code.

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        }
+
+        // * Outer 'for' loop for columns and inner 'for' loops for rows
+        for (int j = 1; j < c; j++)
+        {
+            for (int i = 0; i < r; i++)
+            {
+                if (r == 1)
+                {
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:34]: Remove extra spaces and lines.

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                    // * Only if 1 row
+                    dp[i][j] = mine[i][j] + dp[i][j - 1];
+                }
+                else if (i == 0)
+                {
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:40]: Remove extra spaces and lines.

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+                    // * If it is the first row
+                    dp[i][j] = mine[i][j] + max(dp[i][j - 1], dp[i + 1][j - 1]);
+                }
+                else if (i == r - 1)
+                {
+
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:46]: Remove extra spaces and lines.

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+                    // * If it is last row
+                    dp[i][j] = mine[i][j] + max(dp[i][j - 1], dp[i - 1][j - 1]);
+                }
+                else
+                {
+
+                    // * If it is intermediate row
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:53]: Remove extra spaces and lines.

File: C-Plus-Plus/dp/Goldmine.cpp [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+                    dp[i][j] = mine[i][j] + max(dp[i][j - 1], max(dp[i - 1][j - 1], dp[i + 1][j - 1]));
+                }
+            }
+        }
+
+        // * Returning maximum gold
```
Comment (ankitaggarwal23) [inline C-Plus-Plus/dp/Goldmine.cpp:59]: Remove extra spaces and unnecessary comments.

**Review by ankitaggarwal23 — CHANGES_REQUESTED** [review]
Few changes: 

- Use proper code Format 
- Remove extra lines of code and spaces
- Add sample I/O at the end of the Code
- Add Time and Space Complexity at the end of the Code

**Pull request closed without merging**